### DISCUSSION TITLE
Should DTE be avoided?

### DISCUSSION THREAD

**Discussion by reduckted** [VsixCommunity/Community.VisualStudio.Toolkit#66] [Q&A] [answered]
I've been thinking about this for a while, but it's probably more important now that Visual Studio 2022 is here. Should using the `DTE` (or `DTE2`) be avoided when there are other ways of getting the functionality that you need, even if it's more complicated?

For example, to get all of the projects in the solution, you can use `DTE.Solution.Projects`, or you can use the `IVSHierarchy` API. There's `DTE.Events`, but there's also `Microsoft.VisualStudio.Shell.Events` (I'm not sure if they provide similar functionality though). There are numerous other examples.

I was under the impression that DTE was old, outdated and possibly not as efficient as using the other services because of all the COM overhead. Is that right? Should we try to avoid it as much as possible?

**Comment by madskristensen** [comment, marked as answer]
Yes, the DTE should be avoided, but that has been hard due to the complexity of the alternatives. I'm in the middle of converting some of the events in the toolkit away from the DTE events.

**Reply by yannduran** [reply]
Why do you say that DTE should be avoided? Is that some insider knowledge? Or a direction that the team want to emphasise?

Personally I think the complicated old COM ways should be avoided at all costs as it leads to hard-to-understand, brittle code. The DTE actually gives some protection. It's all of the complicated old COM interop stuff that's outdated, not DTE. DTE is a managed wrapper around all that hard stuff. The DTE method implementations can be changed much more easily than COM interface-based code, as has been clearly demonstrated by VS 2022.

The issue with DTE vs DTE2 is a non-issue really as nobody should be using DTE anymore in 2021. Switching to DTE2 is a very easy fix to implement. I think that more and more COM-based code should be hidden behind something like DTE, and is exactly what the toolkit itself is trying to accomplish.

VS is in the mess that it is precisely because there's too much very old COM-based code that very few people understand and that nobody wants to change, and preventing improving it.

I've heard the 'avoid DTE' argument for years now, but I've never been given any reasonable explanation as to why.

**Reply by madskristensen** [reply]
I use the DTE all the time for the same reason you list. However, it is old, unmaintained, not async, doesn't follow the latest best practices, is missing capabilities. Also, the Visual Studio team has been recommending people not using it for many years now

**Reply by yannduran** [reply]
You're right, it's old. But the COM stuff is **even older** :-P The COM interfaces are set in concrete and are almost impossible to change. DTE gives the opportunity of changing the implementations to whatever is current without breaking its contract. On the other hand, the COM interfaces are like coding knee deep in molasses, or slowly sinking into quicksand.

Until I hear some **_actual reasons_** from **anybody** as to why it should be avoided, I won't be changing my mind about it, nor will I be stopping using it.

If the the team is seriously suggesting we should resort to using the archaic, brittle, hard to understand, even harder to maintain COM interfaces, then I have my doubts about the team. You know what I mean?

**Reply by shaggygi** [reply]
If trying to avoid/hide the DTE, is it good to expose publicly?  Not sure a way around, though.

For example...
https://github.com/VsixCommunity/Community.VisualStudio.Toolkit/blob/64ffb8141f3a3e5215718cafe3ae53c1304fdd6a/src/Community.VisualStudio.Toolkit.Shared/ItemNode.cs#L124

**Reply by yannduran** [reply]
I agree that it should be used **_internally_**. I don't believe that it should be exposed **publicly**, no.

Just my opinion of course, for what it's worth.

**Comment by shaggygi** [comment]
A few pennies regarding `ItemNode`...

IMHO, `ItemNode` is too vague.

At first, I was thinking maybe `HierarchyItem` and let it implement `IVsHierarchyItem`.  However, I'm starting to lean toward `HierarchyNode` and pass in `IVsHierarchyItem` as it is currently.

As mentioned previously, I'm not sure it is wise to expose `DTE` stuff if the goal is to avoid it.  Plus, it would decrease possible breaking changes when future decisions are made on VS Extension model vNext (at least on the direction of using `DTE` and `IVsXXX`).

I noticed in the [IVsHierarchy doc](https://docs.microsoft.com/en-us/dotnet/api/microsoft.visualstudio.shell.interop.ivshierarchy?view=visualstudiosdk-2019), it states `IVsHierarchy` is not meant to be specific to a Solution/Project hierarchy.

> The IVsHierarchy interface is not used only for project hierarchies. For example, the Server Explorer window implements the IVsHierarchy interface to display its hierarchy, which is not a project hierarchy.

Therefore, I wonder if we should create different types of `Nodes` for the Toolkit.  Meaning, `ItemNode` should be named something related to Projects if it going to expose related functionality.  I would not want one class to include everything as that creates maintenance/confusion issues.  Does that mean `ItemNode` becomes `ProjectItemNode` or `ProjectHierarchyNode`?

When it comes to guidance on avoiding `DTE`, use `IVsXXX`, etc., it would be nice to hear from the VS team on the thinking/reasoning as it would help the community provide input to building out this toolkit.  The unknown makes it more difficult to help/support the effort.

I'm still reading up on all this so take it as "thoughts from a newbie" :smile:

**Reply by madskristensen** [reply]
I went with `SolutionItem`

**Reply by yannduran** [reply]
Nice choice, I like it! But does that conflict in any way with Solution Explorer's _Solution Items_ node though?

**Reply by madskristensen** [reply]
It does not. `Solution Item` is just a string name for the solution folder and doesn't have any underlying types associated with it

**Reply by yannduran** [reply]
Ah cool! I just wanted to mention it, just in case.

**Reply by shaggygi** [reply]
In this case, should `NodeType` be renamed to `SolutionItemType` as it is only used with related?

**Comment by madskristensen** [comment]
The toolkit is now completely free from any `EnvDTE` references

**Reply by StevenRasmussen** [reply]
This is great Mads!  I like the idea around the `SolutionItem` but wondering if it makes sense to have separate classes for different types that inherit from `SolutionItem`, ie: `Solution`, `SolutionFolder`, `Project`, `File`.  I feel like at some point we're going to want to add functionality that would only pertain to say a `Solution`.  For example, we may want to expose a collection of `Projects`.  There are numerous other examples (really we just need to look at the equivalent DTE types to see what sort of properties/methods they have).  Thoughts?

**Reply by madskristensen** [reply]
I've had the same thoughts, but the problem is that when you get the selected items, you have to cast it manually to the right type

**Reply by reduckted** [reply]
Amazing work, Mads!

> wondering if it makes sense to have separate classes for different types that inherit from `SolutionItem`

I think that would be great idea, as long as there's functionality that is exclusive to each sub-type. For example `Project` could have `Load()` and `Unload()` methods, whereas `File` could have `Open()`, etc.

**Reply by CZEMacLeod** [reply]
@madskristensen I think since you can do a select case on type, and/or use linq statements like `selecteditems.OfType<File>()`, casting is not really an issue.
@reduckted I definitely think that all type specific functionality should be restricted to the sub-type. There may be actions which could work on many types, such as OpenInEditor - which would work on SDK projects, files etc. but I think these should be implemented by type, possibly with an interface.

**Reply by madskristensen** [reply]
As I see it, the benefit of distinct types is that we can add a `Build()` method only to `Project` and `OpenInEditor` only to `ProjectItem` for example. The downside is that you must cast every time to use those methods. An alternative is what we've got now where you'd use `VS.Build.BuildProject(SolutionItem project)` to build a project. I miss the ease of `myProjectItem.Build()`, but I don't miss the casting that was needed by using the DTE - it's not obvious nor discoverable. 

So, I think it comes down to a tradeoff between convenience and discoverability - or experienced extender vs. beginner. Is there a way to do this that caters to both?

In addition, `PhysicalFolder` and `SolutionFolder` probably don't need their own subtypes. If true, would that be confusing that only some have their own subtype?

**Reply by CZEMacLeod** [reply]
@madskristensen Won't PhysicalFolders have a path property or something, whereas SolutionFolders don't? If so I think they both have their own subtype.
I guess using something like a type testing operator or switch with pattern matching for the selected item as a project kinda solves both problems...
```c#
var solutionItem = await VS.Selection.GetSelectedItemAsync();
if (solutionItem is ProjectItem project) {
    await project.BuildAsync();
}
```
or
```c#
switch(solutionItem)
{
    case ProjectItem project:
        {
            await project.BuildAsync();
            break;
        }
    case FileItem file:
        {
            file.OpenInEditor();
            break;
        }
    case PhysicalFolder folder:
        {
            folder.OpenInExplorer();
            break;
        }
    default:
        {
            await VS.MessageBox.ShowAsync($"Cannot execute magic command on items of type {solutionItem.GetType().Name}");
            break;
        }
}
```

Things like `OpenInEditor`, `OpenInExplorer` and `BuildAsync` could be implemented as extension methods rather than cluttering up the classes themselves.

```c#
    public static class ProjectExtensions
    {
        public static Task BuildAsync(this ProjectItem project, BuildAction action)
        {
            return VS.Build.BuildProjectAsync(project, action);
        }
    }

    public interface IPhysicalItem
    {
        string Path { get; set; }
    }

    public static class PhysicalItemExtensions
    {
        public static void OpenInExplorer(this IPhysicalItem physicalItem)
        {
            ProcessStartInfo si = new ProcessStartInfo
            {
                FileName = "explorer",
                UseShellExecute = true
            };

            FileAttributes attr = File.GetAttributes(physicalItem.Path);
            if (attr.HasFlag(FileAttributes.Directory) {
                si.Arguments = string.Format("/root,{0}", physicalItem.Path);
            }
            else
            {
                var folder = Path.GetDirectoryName(physicalItem.Path);
                var fileName = Path.GetFileName(physicalItem.Path);
                si.Arguments = string.Format("/root,{0} /select,{1}", folder, fileName);
            }
            Process proc = new Process
            {
                StartInfo = si
            };
            proc.Start();
        }
    }
```

You could also add trycast type extension methods to make it 'discoverable'. e.g.
```c#
    public static class SolutionItemExtensions
    {
        public static ProjectItem AsProjectItem(this SolutionItem solutionItem) => solutionItem as ProjectItem;
        public static FileItem AsFileItem(this SolutionItem solutionItem) => solutionItem as FileItem;
        public static PhysicalFolder AsPhysicalFolder(this SolutionItem solutionItem) => solutionItem as PhysicalFolder;
    }
```
Then you can just do
```c#
var solutionItem = await VS.Selection.GetSelectedItemAsync();
await solutionItem.AsProjectItem()?.BuildAsync();
```

**Reply by shaggygi** [reply]
I like @CZEMacLeod approach.  I'm assuming you could then add your own extensions where you might have other custom things to add to solution.

```csharp
public interface IMyOtherItem
{
        ...
}

public static class MyOtherItemExtensions
{
        public static void DoMyThing(this IMyOtherItem otherItem)
        {
        }
}
```

Therefore, in example of checking above..

```csharp
if (solutionItem is MyOtherItem otherItem)
{
    await otherItem.DoMyThing();
}
```

**Reply by madskristensen** [reply]
That makes sense to me. I like it

**Reply by shaggygi** [reply]
If we went with the extensions approach above, what would the `Type` (SolutionItemType) be for external extensions?  Would they always be `Unknown`?  I recall the C# design team working on "extension everything", but not sure when that feature will be added or if it would help in this scenario.

Also, and in addition if it were `Unknown`, how would you find other items in hierarchy?  Using [FindParent](https://github.com/VsixCommunity/Community.VisualStudio.Toolkit/blob/92ffe8f02f54edd89e19a6001a16ad1fab34253a/src/Community.VisualStudio.Toolkit.Shared/Solution/SolutionItem.cs#L265) for example, how would you find item based on your extension item?

**Reply by madskristensen** [reply]
You can always make your own by assigning an arbitrary integer unit to the `Enum`. I forget how I have done it in the past, but something like this:

```csharp
SolutionItemType type = (SolutionItemType)123;
```

However, I don't know if there is a use case for adding additional item to the list. Alternatively, send a PR to the toolkit to add any missing items.

**Reply by madskristensen** [reply]
I've added `Project` and `PhysicalFile` classes. Do they capture the idea correctly?

**Reply by shaggygi** [reply]
I noticed a mention of having interfaces or not.  Should `SolutionItem` be changed to at least an abstract class?  Recall seeing peeps in the .NET review streams recommending abstracts over interfaces.

**Reply by yannduran** [reply]
@madskristensen I always add an `Unknown` with a value of `0` to all of my enums.

@shaggygi For many years Microsoft has been recommending **_composition_** over **_inheritance_**. I guess it wouldn't be the first time MS has flipped on its own advice.

@madskristensen it'd be nice to compile a list of Microsoft's **latest** advice somewhere, either as questions like this arise, or a researched list. This would be a good place for developers to learn about Microsoft's latest recommendations. Online sources (including videos) are often out of date.
